main.cpp

```diff
@@ -0,0 +1,57 @@
+#include <iostream>
+#include <fstream>
+#include <vector>
+
+#define STB_IMAGE_WRITE_IMPLEMENTATION
+#include "stb_image_write.h"
+
+struct Color {
+    uint8_t r;
+    uint8_t b;
+    uint8_t g;
+};
+
+const Color WHITE = { 255, 255, 255 };
+const Color RED = { 255, 0, 0 };
+
+class Image {
+public:
+    // The pixels are stored left-to-right and top-to-bottom.
+    // 0,0 is top left and width -1 , height -1 is bottom right.
+    std::vector<Color> data;
+    const int width;
+    const int height;
+
+    Image(int width, int height):
+        data(width * height, WHITE),
+        width(width),
+        height(height) {}
+
+    void set(int x, int y, Color color) {
```

Comment:
Hier gibt es ein kleines Problem. `color` ist ein by value Parameter. D.h. es wird unten in main hier rein kopiert und dann wird es ein zweites mal in den Vector kopiert. Das kann man besser machen. Entweder man macht unten `data[index] = std::move(color);`, damit wird nur eine Kopie gemacht. Oder man macht oben den Parameter by const ref, also: `const Color& color`. Mir gefällt eher ehrlich gesagt ein ganz anderer Zugang:

```(cpp)
Color& operator()(size_t x, size_t y) {
    return data[x + y * width];
}

const Color& operator()(size_t x, size_t y) const {
    return data[x + y * width];
}
```

Damit definiert man einen Operator der dir einfach die Referenz auf die Zelle in dem Datenvektor zurückgibt. Damit kann die gesetzt werden. Hat man ein const Image, wird man die Methode nicht verwenden können, deshalb ich auch eine zweite Methode definiert, die const ist.

Damit kann man unten nun einfach folgendes schreiben:
```(cpp)
Image image = Image(100, 100);
for(int i = 0; i < 100; i++) {
    image(i, i) = RED;
}
```